content/blog/2018/appswitch/appswitch.md

```diff
@@ -0,0 +1,81 @@
+
+Dependency ordering is deceptively nuanced with a host of interrelated problems.  Ordering individual services requires having the dependency graph of the services so that they can be brought up starting from leaf nodes back to the root nodes.  It is not easy to construct such a graph and keep it updated overtime as interdependencies evolve with the behavior of the application.  Even if the dependency graph is somehow provided, enforcing the ordering itself is not easy.  Simply starting the services in the specified order obviously won’t do.  A service may have started but not yet ready to accept connections.  That is the problem with docker-compose’ `depends-on` tag, for example.
+
+Apart from introducing sufficiently long sleeps between service startups, a common pattern that is often used is to check for readiness of dependencies before starting a service.  In Kubernetes it is typically done as part of the init container of the pod.  However that means that the entire application would be held up until all its dependencies come alive.  Sometimes applications spend several minutes initializing themselves on startup before making their first outbound connection.  Not allowing the service to start adds substantial overhead to startup time.  Also, checking for readiness of dependencies won’t work in case of "multilevel" interdependencies where a service needs to be started before another service but only after a third service.  Likewise, multiple containers of the same pod may have mutual dependencies that cannot be ordered by simply waiting at the init container.
+
+### Example Scenario: IBM WebSphere ND
+
+Let us consider IBM WebSphere ND -- a widely deployed application middleware -- to grok these problems more closely.  It is a fairly complex framework in itself and consists of a central component called deployment manager (`dmgr`) that manages a set of node instances.  It uses UDP to negotiate cluster membership among the nodes and requires that deployment manager is up and operational before any of the node instances can come up and join the cluster.
+
+Why are we talking about a traditional application in the modern cloud-native context?  It turns out that there are significant gains to be had by enabling them to run on Kubernetes and Istio platforms.  Essentially it’s a part of the modernization journey that allows running traditional apps alongside green-field apps on the same modern platform to facilitate interoperation between the two.  As such, running WebSphere ND on Kubernetes requires a whole lot of additional cushion from AppSwitch in the way of the network environment it expects.  For the purpose of this blog however, I'll focus on the dependency ordering requirement.
```

Comment:
on Kubernetes -> on the Kubernetes

Reply:
"requires a whole lof of additional cushion" is not a common expression, I'm not sure what it is trying to say.

This sentence is introducing AppSwitch for the first time in the text, you might want to phrase that a bit more as an introduction. Something like "Coordinating these dependencies is precisely what AppSwitch can help with", or something like it.